gtk3-macros/src/builder_handlers.rs

```diff
@@ -0,0 +1,200 @@
+// Take a look at the license at the top of the repository in the LICENSE file.
+
+use darling::FromMeta;
+use proc_macro::TokenStream;
+use proc_macro2::TokenStream as TokenStream2;
+use quote::{format_ident, quote, quote_spanned};
+use syn::{
+    parse, spanned::Spanned, Attribute, Error, FnArg, ImplItem, ImplItemMethod, ItemImpl, Meta,
+    MetaList, NestedMeta, PatType, Signature, Type,
+};
+
+#[derive(Debug, Default, FromMeta)]
+#[darling(default)]
+pub struct HandlersImplAttributes {
+    get_handler_fn: Option<String>,
+}
+
+#[derive(Debug, Default, FromMeta)]
+#[darling(default)]
+struct HandlerAttributes {
+    name: Option<String>,
+}
+
+#[derive(Debug)]
+struct HandlerInfo {
+    name: String,
+    sig: Signature,
+}
+
+fn generate_handler(info: &HandlerInfo) -> Result<TokenStream2, Error> {
+    let handler_name = &info.name;
+    let arguments: Vec<TokenStream2> = info.sig.inputs
+        .iter()
+        .skip(1)
+        .enumerate()
+        .map(|(index, arg)| {
+            let arg_type = argument_type(arg)?;
+            Ok(quote_spanned! { arg.span() =>
+                match values[#index].get_some() {
+                    Ok(value) => value,
+                    Err(error) => {
+                        glib::g_critical!("builder handler", "Handler {} expects an argument of type {} but received `{:?}`: {}.", #handler_name, stringify!(#arg_type), values[#index], error);
```

Comment:
This should be a panic in one way or another. Returning `None` can easily cause memory safety issues in the C code calling this signal handler.